src/include/migraphx/shape.hpp

```diff
@@ -89,10 +128,30 @@ struct shape
     type_t type() const;
     const std::vector<std::size_t>& lens() const;
     const std::vector<std::size_t>& strides() const;
+
+    /*!
+     * Return the number of elements in the tensor. Multiply the lengths.
+     */
     std::size_t elements() const;
+
+    /*!
+     * Return the number of total bytes used for storage of the tensor data.
+     * Includes subshapes.
+     */
     std::size_t bytes() const;
+
+    /*!
+     * Return the size of the type of the main shape.
+     * Returns 0 if there are subshapes.
+     */
     std::size_t type_size() const;
 
+    const std::vector<dynamic_dimension>& dyn_dims() const;
+
+    std::vector<std::size_t> min_dyn_dims() const;
+    std::vector<std::size_t> max_dyn_dims() const;
+    std::vector<std::size_t> opt_dyn_dims() const;
```

Comment:
These should probably be `max_lens`, `min_lens`, and `opt_lens` and they should return `lens()` when the shape is not dynamic.